File: maze.js
var cols = 12
var rows = 12;
var w; // Cell size
var maze;

var seed = 513;

function setup() {
  //Good seeds: 126, 513
  //bad seeds 512
  randomSeed(513);

  createCanvas(400, 400);
  w = min(400/cols, 400/rows);

  // Create maze object and initialize grid
  maze = new Maze(cols, rows);
  maze.createGrid();

  var gui = createGui('My awesome GUI');
  gui.addGlobals('seed', 'cols', 'rows');

  //button = createButton("Click me!");
  //button.mousePressed(buttonClicked);
}

function buttonClicked() {
  console.log("Button clicked!");
}

function draw() {
  background(51);

  // Draw all cells
  maze.showGrid();

  // Step the maze generation process
  if (!maze.isComplete) {
    maze.generateStep();
  }
}

// Maze object containing grid, current cell, and stack
class Maze {
  constructor(cols, rows) {
    this.cols = cols;
    this.rows = rows;
    this.grid = [];
    this.current = null;
    this.stack = [];
    this.isComplete = false;
  }

  // Create grid of cells
  createGrid() {
    this.grid = [];
    for (let j = 0; j < this.rows; j++) {
      for (let i = 0; i < this.cols; i++) {
        let cell = new Cell(i, j);
        this.grid.push(cell);
      }
    }
    this.current = this.grid[0];
    this.isComplete = false;
  }

  // Helper to get grid index
  index(i, j) {
    if (i < 0 || j < 0 || i >= this.cols || j >= this.rows) {
      return -1;
    }
    return i + j * this.cols;
  }

  // Draw all cells in the grid
  showGrid() {
    for (let i = 0; i < this.grid.length; i++) {
      this.grid[i].show();
    }

    if (this.current) {
      this.current.highlight();
    }
  }

  // Single step of maze generation
  generateStep() {
    if (this.isComplete) {
      return false;
    }

    this.current.visited = true;

    // Step 1: Choose a random neighbor
    let next = this.current.checkNeighbors(this);
    if (next) {
      next.visited = true;

      // Step 2: Push current cell to the stack
      this.stack.push(this.current);

      // Step 3: Remove the wall between current and next
      this.removeWalls(this.current, next);

      // Step 4: Make the chosen cell the current cell
      this.current = next;
    } else if (this.stack.length > 0) {
      this.current = this.stack.pop();
    } else {
      // Maze generation is complete
      this.isComplete = true;
    }
    return true;
  }

  // Remove walls between two cells
  removeWalls(a, b) {
    let x = a.i - b.i;
    if (x === 1) {
      a.walls[3] = false;
      b.walls[1] = false;
    } else if (x === -1) {
      a.walls[1] = false;
      b.walls[3] = false;
    }

    let y = a.j - b.j;
    if (y === 1) {
      a.walls[0] = false;
      b.walls[2] = false;
    } else if (y === -1) {
      a.walls[2] = false;
      b.walls[0] = false;
    }
  }
}

// Cell object
class Cell {
  constructor(i, j) {
    this.i = i;
    this.j = j;
    this.walls = [true, true, true, true]; // Top, right, bottom, left
    this.visited = false;
  }

  show() {
    let x = this.i * w;
    let y = this.j * w;
    stroke(255);
    if (this.walls[0]) line(x, y, x + w, y); // Top
    if (this.walls[1]) line(x + w, y, x + w, y + w); // Right
    if (this.walls[2]) line(x + w, y + w, x, y + w); // Bottom
    if (this.walls[3]) line(x, y + w, x, y); // Left

    if (this.visited) {
      noStroke();
      fill(255, 0, 255, 100);
      rect(x, y, w, w);
    }
  }

  highlight() {
    let x = this.i * w;
    let y = this.j * w;
    noStroke();
    fill(0, 255, 0, 150);
    rect(x, y, w, w);
  }

  checkNeighbors(maze) {
    let neighbors = [];

    let top = maze.grid[maze.index(this.i, this.j - 1)];
    let right = maze.grid[maze.index(this.i + 1, this.j)];
    let bottom = maze.grid[maze.index(this.i, this.j + 1)];
    let left = maze.grid[maze.index(this.i - 1, this.j)];

    if (top && !top.visited) neighbors.push(top);
    if (right && !right.visited) neighbors.push(right);
    if (bottom && !bottom.visited) neighbors.push(bottom);
    if (left && !left.visited) neighbors.push(left);

    if (neighbors.length > 0) {
      let r = floor(random(0, neighbors.length));
      return neighbors[r];
    } else {
      return undefined;
    }
  }
}

function download(data, filename, type) {
    var file = new Blob([data], {type: type});
    if (window.navigator.msSaveOrOpenBlob) // IE10+
        window.navigator.msSaveOrOpenBlob(file, filename);
    else { // Others
        var a = document.createElement("a"),
                url = URL.createObjectURL(file);
        a.href = url;
        a.download = filename;
        document.body.appendChild(a);
        a.click();
        setTimeout(function() {
            document.body.removeChild(a);
            window.URL.revokeObjectURL(url);
        }, 0);
    }
}

File: old_prototypes/maze.py
import json
import random
from dataclasses import dataclass, field

import pygame

# Constants
WIDTH, HEIGHT = 400, 400
COLS, ROWS = 12, 12
W = min(WIDTH // COLS, HEIGHT // ROWS)  # Cell size

# Colors
BLACK = (0, 0, 0)
WHITE = (255, 255, 255)
MAGENTA = (255, 0, 255, 100)
GREEN = (0, 255, 0, 150)

# Initialize Pygame
pygame.init()
screen = pygame.display.set_mode((WIDTH, HEIGHT))
clock = pygame.time.Clock()

#Seed
seed = random.randrange(0, 2**32, 1)
print("Using seed: ", seed)
random.seed(seed)

# Grid
grid = []
stack = []
current = None

def reset_state():
    seed = random.randrange(0, 2**32, 1)
    print("Using seed: ", seed)
    random.seed(seed)
    grid = []
    stack = []
    current = None


# Helper to get grid index
def index(i, j):
    if i < 0 or j < 0 or i >= COLS or j >= ROWS:
        return -1
    return i + j * COLS

# Cell class
@dataclass
class Cell:
    i: int
    j: int
    walls: list[bool] = field(default_factory = lambda: list((True, True, True, True))) #self.walls = [True, True, True, True]  # Top, right, bottom, left
    visited: bool = False

    # def __init__(self, i, j):
    #     self.i = i
    #     self.j = j
    #     self.walls = [True, True, True, True]  # Top, right, bottom, left
    #     self.visited = False

    def show(self):
        x, y = self.i * W, self.j * W
        if self.visited:
            pygame.draw.rect(screen, MAGENTA, (x, y, W, W))
        if self.walls[0]:
            pygame.draw.line(screen, WHITE, (x, y), (x + W, y))  # Top
        if self.walls[1]:
            pygame.draw.line(screen, WHITE, (x + W, y), (x + W, y + W))  # Right
        if self.walls[2]:
            pygame.draw.line(screen, WHITE, (x + W, y + W), (x, y + W))  # Bottom
        if self.walls[3]:
            pygame.draw.line(screen, WHITE, (x, y + W), (x, y))  # Left

    def highlight(self):
        x, y = self.i * W, self.j * W
        pygame.draw.rect(screen, GREEN, (x, y, W, W))

    def check_neighbors(self):
        neighbors = []

        top = grid[index(self.i, self.j - 1)] if index(self.i, self.j - 1) != -1 else None
        right = grid[index(self.i + 1, self.j)] if index(self.i + 1, self.j) != -1 else None
        bottom = grid[index(self.i, self.j + 1)] if index(self.i, self.j + 1) != -1 else None
        left = grid[index(self.i - 1, self.j)] if index(self.i - 1, self.j) != -1 else None

        if top and not top.visited:
            neighbors.append(top)
        if right and not right.visited:
            neighbors.append(right)
        if bottom and not bottom.visited:
            neighbors.append(bottom)
        if left and not left.visited:
            neighbors.append(left)

        if neighbors:
            return random.choice(neighbors)
        else:
            return None

# Remove walls between two cells
def remove_walls(a, b):
    dx = a.i - b.i
    if dx == 1:
        a.walls[3] = False
        b.walls[1] = False
    elif dx == -1:
        a.walls[1] = False
        b.walls[3] = False

    dy = a.j - b.j
    if dy == 1:
        a.walls[0] = False
        b.walls[2] = False
    elif dy == -1:
        a.walls[2] = False
        b.walls[0] = False

# Create grid
for j in range(ROWS):
    for i in range(COLS):
        grid.append(Cell(i, j))

current = grid[0]

# Main loop
running = True
while running:
    screen.fill(BLACK)

    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False

    # Draw grid
    for cell in grid:
        cell.show()

    # Maze generation logic
    current.visited = True
    current.highlight()

    next_cell = current.check_neighbors()
    if next_cell:
        next_cell.visited = True
        stack.append(current)
        remove_walls(current, next_cell)
        current = next_cell
    elif stack:
        current = stack.pop()

    keys = pygame.key.get_pressed()
    if keys[pygame.K_s]:
        fileName = "maze-%d.json" % seed
        with open(fileName, "w") as fh:
            json.dump([c.__dict__ for c in grid], fh)
        print("Maze dumped to: ", fileName)
    if keys[pygame.K_q]:
        print("Exiting.")
        running = False
    if keys[pygame.K_r]:
        print("Resetting maze.")
        reset_state()


    pygame.display.flip()
    clock.tick(30)

pygame.quit()
